node/rate_limiter.js

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_SERVICE_RPS_LIMIT = 100;
+var DEFAULT_TOTAL_RPS_LIMIT = 1000;
+var DEFAULT_BUCKET_NUMBER = 20;
+
+function RateLimiterCounter(options) {
+    if (!(this instanceof RateLimiterCounter)) {
+        return new RateLimiterCounter(options);
+    }
+
+    var self = this;
+    self.index = 0;
+    self.rps = 0;
+    self.numOfBuckets = options.numOfBuckets;
+    self.buckets = [];
+    self.buckets.length = self.numOfBuckets;
```

Comment:
This line makes V8 turn the array implementation into a hashmap. Does not matter that you fill it in in one second. V8 will have to convert between hashmap and flat array behind the scenes between when you execute this line and one second later, if it even has the logic to detect going from sparse back to compact. Delete this line and it will just go fast. It does not help.